File: login-server/src/main/java/work/fking/pangya/login/auth/Authenticator.java
package work.fking.pangya.login.auth;

import work.fking.pangya.common.Rand;

public interface Authenticator {

    Authenticator NOOP_AUTHENTICATOR = (username, password) -> new UserInfo(Rand.max(10000), username, username);

    UserInfo authenticate(String username, char[] password) throws Exception;
}

File: login-server/src/main/kotlin/work/fking/pangya/login/auth/Authenticator.kt
package work.fking.pangya.login.auth


val NOOP_AUTHENTICATOR: Authenticator = Authenticator { _, _ -> InvalidCredentialsResult }

fun interface Authenticator {
    fun authenticate(username: String, password: CharArray): AuthResult
}

sealed interface AuthResult
data class SuccessResult(val userInfo: UserInfo) : AuthResult
data class ExceptionResult(val exception: Exception) : AuthResult
data object InvalidCredentialsResult : AuthResult
